tools/unit/src/main/kotlin/com/datadog/tools/unit/extensions/TestConfigurationExtension.kt

```diff
@@ -0,0 +1,110 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.tools.unit.extensions
+
+import android.annotation.SuppressLint
+import com.datadog.tools.unit.annotations.TestConfigurationsProvider
+import com.datadog.tools.unit.extensions.config.TestConfiguration
+import fr.xgouchet.elmyr.Forge
+import fr.xgouchet.elmyr.junit5.ForgeExtension
+import org.junit.jupiter.api.extension.AfterAllCallback
+import org.junit.jupiter.api.extension.AfterEachCallback
+import org.junit.jupiter.api.extension.BeforeAllCallback
+import org.junit.jupiter.api.extension.BeforeEachCallback
+import org.junit.jupiter.api.extension.ExtensionContext
+
+/**
+ * A JUnit Jupiter extension that can ensure a test configuration state is properly handled with
+ * self contained independent configurations.
+ */
+class TestConfigurationExtension :
+    BeforeAllCallback,
+    BeforeEachCallback,
+    AfterEachCallback,
+    AfterAllCallback {
+
+    private val testConfigurations: MutableMap<String, List<TestConfiguration>> = mutableMapOf()
+
+    // region BeforeAllCallback
+
+    /** @inheritdoc */
+    override fun beforeAll(context: ExtensionContext?) {
+        checkNotNull(context)
+        val methods = context.requiredTestClass
```

Comment:
it took me a while to understand the logic here and how it connects with the test class. Could you please add a small explanation here ? Also could you mention how an `ExtensionContext` relates to its parent maybe by giving an example ?

Reply:
Quoting from JUnit 5 doc:
> ExtensionContext encapsulates the context in which the current test or container is being executed.

A test method's context's parent will be the test class's context (and you can have nested contexts in case of test methods running more than once, with `@Repeatable` or `@ParameterizedTest` for instance)